### PR TITLE
Rnd 770 | configure redux for using multiple tabs

### PR DESCRIPTION
**Pull request by SzajekSzymon**
## Description

provide a logic and redux store structure to make possible using extension for a multiple tabs in chrome.

## Motivation and Context

In previous version opening devtool for next tab was causing an overwrite data in store. 
As a result all open devtools with knotx etension geted the same data for only one ( current active) tab

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: src/js/devtools/App.jsx [hunk 1]
```diff
@@ -17,9 +17,9 @@
 import React from 'react';
 import ExampleGraph from '../components/Graphs/exampleComponent';
 
-const App = () => (
+const App = (props) => (
   <div className="app">
-    <ExampleGraph />
+    <ExampleGraph {...props} />
```
Comment (xolir) [inline src/js/devtools/App.jsx]: There is only one thing being passed via this object, so it's better to pass it directly

File: .eslintrc.json [hunk 2]
```diff
@@ -26,6 +26,7 @@
     ],
     "rules": {
         "max-len": ["error", { "code": 120 }],
-        "import/prefer-default-export": "off"
+        "import/prefer-default-export": "off",
+        "react/jsx-props-no-spreading": "off"
```
Comment (xolir) [inline .eslintrc.json]: I don't think this should be removed

File: src/js/components/Graphs/exampleComponent.jsx [hunk 3]
```diff
@@ -21,8 +21,10 @@ import vis from 'vis-network';
 import { connect } from 'react-redux';
 import PropTypes from 'prop-types';
 import { bindActionCreators } from 'redux';
+// import { store } from '../../state/store';
```
Comment (xolir) [inline src/js/components/Graphs/exampleComponent.jsx]: Is this needed now?

File: src/js/components/Graphs/exampleComponent.jsx [hunk 4]
```diff
@@ -85,17 +89,18 @@ const ExampleGraph = ({
...
   actions: PropTypes.shape({
-    getPageData: PropTypes.func.isRequired,
+    setPageData: PropTypes.func.isRequired,
   }).isRequired,
+  tabId: PropTypes.number.isRequired,
 };
 
-const mapStateToProps = ({ pageData }) => ({
-  pageUrl: pageData.url,
+const mapStateToProps = ({ pageData }, onwProps) => ({
```
Comment (xolir) [inline src/js/components/Graphs/exampleComponent.jsx]: `onwProps` -> `ownProps `

File: src/js/state/reducers/pageData.spec.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+
+import pageDataReducer, { initState } from './pageData';
+import { SET_PAGE_DATA, REMOVE_PAGE_DATA } from '../actionTypes/pageData';
+
+
+const examplePadDatajObj = [
```
Comment (xolir) [inline src/js/state/reducers/pageData.spec.js]: Some mispelling here

File: src/js/state/reducers/pageData.spec.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
+
+import pageDataReducer, { initState } from './pageData';
+import { SET_PAGE_DATA, REMOVE_PAGE_DATA } from '../actionTypes/pageData';
+
+
+const examplePageDataObj = [
```
Comment (tomaszmichalak) [inline src/js/state/reducers/pageData.spec.js]: Maybe something like:
```
const pageData = {
  id: 'example_id_0',
  url: 'www.exampleurl0.com',
  fragments: []
}

const nextPageData = {
  id: 'example_id_1',
  url: 'www.exampleurl1.com',
  fragments: []
}

const allPages = [
  pageData,
  nextPageData
];
```
Then you can model expected models using references to pageData and nextPageData.

File: src/js/state/reducers/pageData.spec.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+describe('pageData reducer', () => {
+  test('Handles empty action', () => {
+    const state = pageDataReducer(undefined, {});
+    expect(state).toEqual(initState);
+  });
+
+  // SET_PAGE_DATA
+  test('Handles SET_PAGE_DATA event with initial state', () => {
+    const expectedData = {
+      example_id_0: {
```
Comment (tomaszmichalak) [inline src/js/state/reducers/pageData.spec.js]: Now you can simply reference to previous objects:
```
const expectedData = {
  [pageData.id]: pageData.url
}
```

File: src/js/background/background.js [hunk 8]
```diff
@@ -16,37 +16,71 @@
...
+chrome.runtime.onConnect.addListener((port) => {
+  if (port.name !== KNOTX_DEVTOOL_CONNECTION) return;
+
+  port.onMessage.addListener((requestType) => {
+    if (requestType === GET_CURRENT_TAB_INFO) {
+      chrome.tabs.query({
+        active: true,
+      },
+      (tabs) => {
+        port.postMessage(tabs[0]);
```
[the remark this answers is not in the document]
Comment (SzajekSzymon) [reply src/js/background/background.js]: I can't. This is chrome API method. This methods send message in scope of a port (connection), and I provide only current tab (tabs[0])

File: src/js/background/background.js [hunk 9]
```diff
@@ -16,37 +16,71 @@
...
+      id: sender.tab.id,
       url: sender.tab.url,
     };
 
-    store.dispatch(getPageData(pageDataObj));
+    store.dispatch(setPageData(pageDataObj));
 
     const { pageData } = store.getState();
-    if (pageData.fragments.length > 0 && pageData.url) {
+    if (pageData[pageDataObj.id].fragments.length > 0 && pageData[pageDataObj.id].url) {
```
[the remark this answers is not in the document]
Comment (SzajekSzymon) [reply src/js/background/background.js]: in the scope of one process ( in scope of background ) store is synchronous

File: src/js/components/Graphs/exampleComponent.jsx [hunk 10]
```diff
@@ -65,6 +66,7 @@ const exampleGraphDeclaration = {
 
 const ExampleGraph = ({
   pageUrl,
+  tabId,
   // actions,
```
Comment (mikolajkozakiewicz) [inline src/js/components/Graphs/exampleComponent.jsx]: Is this needed?

File: src/js/index.jsx [hunk 11]
```diff
@@ -18,19 +18,29 @@ import React from 'react';
 import ReactDOM from 'react-dom';
 import { Store } from 'webext-redux';
 import { Provider } from 'react-redux';
-import App from './apps/App';
+import App from './devtools/App';
 import '../sass/style.scss';
-import { REDUX_PORT, PANEL_NAME } from './helpers/constants';
-
-const store = new Store({ portName: REDUX_PORT });
+import {
```
Comment (mikolajkozakiewicz) [inline src/js/index.jsx]: I would consider adding some scoping here, for example place GET_CURRENT_TAB_INFO in messages object

File: src/js/state/reducers/pageData.spec.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+
+describe('pageData reducer', () => {
+  test('Handles empty action', () => {
+    const state = pageDataReducer(undefined, {});
+    expect(state).toEqual(initState);
+  });
+
+
+  // SET_PAGE_DATA
```
Comment (mikolajkozakiewicz) [inline src/js/state/reducers/pageData.spec.js]: This comment is redundant with test description

File: src/js/state/reducers/pageData.spec.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        fragments: nextPageData.fragments,
+      },
+    };
+
+    const state = pageDataReducer(initialState, { type: SET_PAGE_DATA, pageData: nextPageData });
+    expect(state).toEqual(expectedData);
+  });
+
+
+  // REMOVE_PAGE_DATA
```
Comment (mikolajkozakiewicz) [inline src/js/state/reducers/pageData.spec.js]: This comment is redundant aswell

File: src/js/state/reducers/pageData.spec.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+  test('Handles REMOVE_PAGE_DATA event with initial state', () => {
+    const state = pageDataReducer(undefined, { type: REMOVE_PAGE_DATA, pageData: { id: currentPageData.id } });
+    expect(state).toEqual(initState);
+  });
+
+  test('Handles REMOVE_PAGE_DATA event with some data already exitsed in state', () => {
```
Comment (mikolajkozakiewicz) [inline src/js/state/reducers/pageData.spec.js]: `exitsed` -> `existing`

File: src/js/state/reducers/pageData.spec.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+      },
+    };
+
+    const state = pageDataReducer(initialState, { type: SET_PAGE_DATA, pageData: nextPageData });
+    expect(state).toEqual(expectedData);
+  });
+
+
+  // REMOVE_PAGE_DATA
+  test('Handles REMOVE_PAGE_DATA event with initial state', () => {
```
Comment (mikolajkozakiewicz) [inline src/js/state/reducers/pageData.spec.js]: Also i would change all occurences of 'event' to 'action' to be more redux-specific

File: src/js/components/Graphs/exampleComponent.jsx (hunk 10 above)
Comment (SzajekSzymon) [reply src/js/components/Graphs/exampleComponent.jsx]: this is example component. I would like to show how to use actions for those who never use redux before. Example component will be remove in the future

**Pull request merged**